silq/analysis/analysis.py

```diff
@@ -715,8 +717,85 @@ def analyse_EPR(empty_traces: np.ndarray,
             'mean_high_blip_duration': results_read['mean_high_blip_duration']}
 
 
-def analyse_flips(up_proportions_arrs: List[np.ndarray],
-                  threshold_up_proportion: Union[Sequence, float]):
+def determine_threshold_up_proportion_single_state(up_proportions_arr: np.ndarray,
+                                                   shots_per_frequency: int):
+    """ Determine threshold up-proportion for single nuclear state readout using
+        single up proportion array of samples
+
+    Using kernel density estimation, 'determine_threshold_up_proportion_single_state'
+    function determines the probability density distribution of measured electron spin-up
+    proportions within the 0 to 1 up proportion space in the form of a Gaussian function.
+    It is later analysed for the peaks in the density distribution and based on their
+    relative position within 0 to 1 up proportion space, the density minimum is found,
+    which is considered to be the threshold for determining the nuclear spin state.
+
+    Args:
+        up_proportions_arr: 1D Array of up proportions, calculated from
+            `analyse_traces`. Up proportion is the proportion of traces that
+            contain blips. The length is number of samples.
+        shots_per_frequency: Integer number of ESR shots/traces, which are
+            used to determine the up proportion.
+
+    Returns:
+        (float) threshold up-proportion
+    """
+    assert isinstance(up_proportions_arr, np.ndarray), 'Up proportions must be a 1D array, not a list.'
+
+    num_shots = max(shots_per_frequency, 25)
+    logger.warning("Number of shots is < 25, 25 shots is used for threshold analysis.")
```

Comment:
Warning is a bit vague atm. What about:
`logger.warning(f'Readout shots {shots_per_frequency} < 25. This may cause issues when finding an optimal threshold up proportion.')

Reply:
corrected.